pulsar-broker/src/test/java/org/apache/pulsar/client/api/TokenExpirationProduceConsumerTest.java

```diff
@@ -0,0 +1,169 @@
+        conf.setBrokerClientAuthenticationPlugin(AuthenticationToken.class.getName());
+        conf.setBrokerClientAuthenticationParameters("token:" + ADMIN_TOKEN);
+        conf.getProperties().setProperty("tokenSecretKey", "data:;base64,"
+                + Base64.getEncoder().encodeToString(SECRET_KEY.getEncoded()));
+    }
+
+    private PulsarClient getClient(String token) throws Exception {
+        ClientBuilder clientBuilder = PulsarClient.builder()
+                .serviceUrl(pulsar.getBrokerServiceUrlTls())
+                .tlsTrustCertsFilePath(TLS_TRUST_CERT_FILE_PATH)
+                .enableTls(true)
+                .allowTlsInsecureConnection(false)
+                .enableTlsHostnameVerification(true)
+                .authentication(AuthenticationToken.class.getName(),"token:" +token)
+                .operationTimeout(1000, TimeUnit.MILLISECONDS);
+        return clientBuilder.build();
+    }
+
+    private PulsarAdmin getAdmin(String token) throws Exception {
+        PulsarAdminBuilder clientBuilder = PulsarAdmin.builder().serviceHttpUrl(pulsar.getWebServiceAddressTls())
+                .tlsTrustCertsFilePath(TLS_TRUST_CERT_FILE_PATH)
+                .allowTlsInsecureConnection(false)
+                .authentication(AuthenticationToken.class.getName(),"token:" +token)
+                .enableTlsHostnameVerification(true);
+        return clientBuilder.build();
+    }
+
+    @Test(timeOut = 60 * 1000)
+    public void testTokenExpirationProduceConsumer() throws Exception {
+        Calendar calendar = Calendar.getInstance();
+        calendar.add(Calendar.SECOND, 20);
+        String role = "test";
+        String token = getExpireToken(role, calendar.getTime());
+        Date expiredTime = calendar.getTime();
+
+        Set<AuthAction> permissions = new HashSet<>();
+        permissions.add(AuthAction.consume);
+        permissions.add(AuthAction.produce);
+        admin.namespaces().grantPermissionOnNamespace(namespaceName.toString(), role, permissions);
+
+        @Cleanup
+        PulsarClient pulsarClient = getClient(token);
+        String topic = namespaceName + "/test-token";
+
+        @Cleanup final Consumer<byte[]> consumer = pulsarClient.newConsumer()
+                .topic(topic)
+                .subscriptionName("test-token")
+                .subscribe();
+        @Cleanup final Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topic)
+                .create();
+        while (true) {
+            try {
+                producer.send("heart beat".getBytes());
+                Message<byte[]> message = consumer.receive();
+                consumer.acknowledge(message);
+                TimeUnit.SECONDS.sleep(3);
+            } catch (Exception e) {
+                assertTrue(new Date().compareTo(expiredTime) > 0
+                        && e instanceof PulsarClientException.TimeoutException);
```

Comment:
@codelipenghui  Thanks for your review.

> Looks the issue is introduced by https://github.com/apache/pulsar/pull/14044?

Yes.

> This will confuse users by getting a TimeoutException after the token gets expires.

The Producer can not send message, it will throw a send timeout exception.